mgr/rbd_support: remove localized schedule option during module startup

ajarr opened the pull request:
```
The localized schedule option was removed during every load of a
scheduler handler. This occured multiple times during a mgr process's
lifespan. Instead, remove the localized schedule module option when the
rbd_support module starts up.

Fixes: https://tracker.ceph.com/issues/57726
Signed-off-by: Ramana Raja <rraja@redhat.com>
```

For more context, you can see the discussion here, https://github.com/ceph/ceph/pull/37864/commits/8e9e2a73f5b8abd743933e59a3b986598765a681#r513533495

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

ajarr commented:
I've not tested this fix yet. Documenting progress.

trociny commented on src/pybind/mgr/rbd_support/schedule.py:
```diff
@@ -387,17 +387,18 @@ def load(self,
         schedule_cfg = self.handler.module.get_module_option(
             self.handler.MODULE_OPTION_NAME, '')
+        schedule_local_cfg = self.handler.module.get_localized_module_option(
+             self.handler.MODULE_OPTION_NAME, '')
```
I would move this below the comment below, as it is part of that and we will remove that whole block with the comment  in some future.

ajarr replied:
Done

trociny reviewed (commented):
Otherwise LGTM

trociny commented on src/pybind/mgr/rbd_support/schedule.py:
```diff
@@ -387,17 +387,18 @@ def load(self,
         schedule_cfg = self.handler.module.get_module_option(
             self.handler.MODULE_OPTION_NAME, '')
+        schedule_local_cfg = self.handler.module.get_localized_module_option(
+             self.handler.MODULE_OPTION_NAME, '')
```
nit: is indentation correct here? In github it looks like there is an extra space.

idryomov replied:
Yup, here is flake8 complaint:
```
rbd_support/schedule.py:391:14: E126 continuation line over-indented for hanging indent
1     E126 continuation line over-indented for hanging indent
```

trociny commented on src/pybind/mgr/rbd_support/schedule.py:
```diff
@@ -387,17 +387,18 @@ def load(self,
+            if not schedule_cfg:
+                schedule_cfg = schedule_local_cfg
                 self.handler.module.set_module_option(
                     self.handler.MODULE_OPTION_NAME, schedule_cfg)
```
nit: It's really highly unlikely, but still there is a chance the mgr crashes between removing the localized option and setting the global option. So I would prefer we removed the localized option only after setting the global.

ajarr replied:
good point. done.

ajarr commented:
@trociny I've added you as the co-author of the PR since my PR ended up being identical to yours after addressing your comments. I hope that's OK with you.

idryomov commented:
jenkins test make check

ajarr commented:
I think I found a bug when testing. Converting this to draft. Will update tomorrow once I confirm.

ajarr commented:
Testing this PR with vstart cluster:
- Set the global snapshot schedule option
```
$ rbd mirror snapshot schedule add 3m
```
- Checked the mgr/rbd_support/mirror_snapshot_schedule config
```
$ ceph config get mgr mgr/rbd_support/mirror_snapshot_schedule
[
    {
        "interval": "3m",
        "start_time": null
    }
]
```

- Observed the audit log
```
2022-12-10T13:39:25.918713-0500 mon.a (mon.0) 208 : audit [INF] from='mgr.4110 10.0.0.148:0/77460' entity='mgr.x' cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"}]: dispatch
2022-12-10T13:40:25.940753-0500 mon.a (mon.0) 209 : audit [INF] from='mgr.4110 10.0.0.148:0/77460' entity='mgr.x' cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"}]: dispatch
2022-12-10T13:41:25.962968-0500 mon.a (mon.0) 210 : audit [INF] from='mgr.4110 10.0.0.148:0/77460' entity='mgr.x' cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"}]: dispatch
2022-12-10T13:42:25.984646-0500 mon.a (mon.0) 211 : audit [INF] from='mgr.4110 10.0.0.148:0/77460' entity='mgr.x' cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"}]: dispatch
2022-12-10T13:43:26.006289-0500 mon.a (mon.0) 212 : audit [INF] from='mgr.4110 10.0.0.148:0/77460' entity='mgr.x' cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"}]: dispatch
2022-12-10T13:44:26.028035-0500 mon.a (mon.0) 213 : audit [INF] from='mgr.4110 10.0.0.148:0/77460' entity='mgr.x' cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"}]: dispatch
2022-12-10T13:45:26.050692-0500 mon.a (mon.0) 214 : audit [INF] from='mgr.4110 10.0.0.148:0/77460' entity='mgr.x' cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"}]: dispatch
2022-12-10T13:46:26.067157-0500 mon.a (mon.0) 215 : audit [INF] from='mgr.4110 10.0.0.148:0/77460' entity='mgr.x' cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"}]: dispatch
2022-12-10T13:47:26.088667-0500 mon.a (mon.0) 217 : audit [INF] from='mgr.4110 10.0.0.148:0/77460' entity='mgr.x' cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"}]: dispatch
```
Saw repeated `config rm mgr/rbd_support/x/mirror_snapshot_schedule` being issued. I expected this PR to stop this command from being issue repeatedly.

- Looked into the mgr.x log
```
2022-12-10T13:49:26.133-0500 7f60c5c5c640  0 [rbd_support DEBUG root] MirrorSnapshotScheduleHandler: refresh_images
2022-12-10T13:49:26.133-0500 7f60c5c5c640  0 [rbd_support INFO root] MirrorSnapshotScheduleHandler: load_schedules
2022-12-10T13:49:26.133-0500 7f60c5c5c640 20 mgr get_config  key: mgr/rbd_support/mirror_snapshot_schedule
2022-12-10T13:49:26.133-0500 7f60c5c5c640 10 mgr get_typed_config get_typed_config mirror_snapshot_schedule found
2022-12-10T13:49:26.133-0500 7f60c5c5c640 20 mgr get_config  key: mgr/rbd_support/x/mirror_snapshot_schedule
2022-12-10T13:49:26.133-0500 7f60c5c5c640 20 mgr get_config  key: mgr/rbd_support/mirror_snapshot_schedule
2022-12-10T13:49:26.133-0500 7f60c5c5c640 10 mgr get_typed_config get_typed_config mirror_snapshot_schedule found
2022-12-10T13:49:26.133-0500 7f60c5c5c640 10 start_mon_command cmd=[{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"}]
2022-12-10T13:49:26.133-0500 7f60c5c5c640 10 monclient: _send_command 82 [{"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"}]
2022-12-10T13:49:26.133-0500 7f60c5c5c640 10 monclient: _send_mon_message to mon.a at v2:10.0.0.148:40903/0
2022-12-10T13:49:26.133-0500 7f60c5c5c640  1 -- 10.0.0.148:0/77460 --> [v2:10.0.0.148:40903/0,v1:10.0.0.148:40904/0] -- mon_command({"prefix":"config rm","who":"mgr","name":"mgr/rbd_support/x/mirror_snapshot_schedule"} v 0) v1 -- 0x560232f3f980 con 0x560232354400
```
It looks `get_localized_module_option()` fetches the global "rbd_support" module option "mirror_snapshot_schedule" if the local option is not present. I expected `get_localized_module_option()` to return an empty string if the localized option is not available. Digging further into the mgr code,  `get_localized_module_option()` calls `PyObject *ActivePyModules::get_typed_config()`, which fetches the global module option if the global option is not present.
See, https://github.com/ceph/ceph/blame/c2737576011bdf333477ff6b6e343a11a5db396b/src/mgr/ActivePyModules.cc#L684

trociny commented on src/pybind/mgr/rbd_support/schedule.py:
```diff
@@ -390,14 +390,15 @@ def load(self,
-        if not schedule_cfg:
-            schedule_cfg = self.handler.module.get_localized_module_option(
-                self.handler.MODULE_OPTION_NAME, '')
-            if schedule_cfg:
```
I suppose moving the code below under `if not schedule_cfg:` block should fix the issue you described.

ajarr replied:
I thought of this too. But I don't think this will result in the local module option being detected and removed when both the global module option and the local module option exist. You described this situation here,
https://github.com/ceph/ceph/pull/37864#discussion_r513533495
_I intentionally put this out of the if clause, so it always had a chance to clean the old option. Consider a case when an old localized option exists for both mgr.x and mgr.y (mgr/rbd_support/x/mirror_snapshot_schedule and mgr/rbd_support/y/mirror_snapshot_schedule). After upgrade the active mgr.x "fixes" mgr/rbd_support/x/mirror_snapshot_schedule making it mgr/rbd_support/mirror_snapshot_schedule. Then mgr.y becomes active, and if it is under if clause it will not remove mgr/rbd_support/y/mirror_snapshot_schedule because mgr/rbd_support/mirror_snapshot_schedule exists. And then if the user decides to remove the global schedule (mgr/rbd_support/mirror_snapshot_schedule) the mgr.y will mysteriously reuse mgr/rbd_support/y/mirror_snapshot_schedule_

trociny replied:
Indeed. Still, taking how unlikely one will face this situation now after we have been have this "remove localized option" code for more than two years (and actually about to remove it as unnecessary any more), especially taking into the consideration that the bug was introduced at the time when the feature was new and not many users used it,  I vote for this `if not schedule_cfg`  solution, unless someone see a better one. Fixing the audit log flood looks more important to me.

idryomov replied:
I second the "do it on init" approach.  Then `load` could always query just the global option, making the common code path more straightforward.

ajarr replied:
Done. Thanks!

idryomov commented on src/pybind/mgr/rbd_support/module.py:
```diff
@@ -80,6 +80,21 @@ def __init__(self, *args: Any, **kwargs: Any) -> None:
+        # Previous scheduler handler versions incorrectly stored the
+        # global config in the localized module option. Check the config is
+        # here and fix it.
+        for handler in (self.mirror_snapshot_schedule, self.trash_purge_schedule):
```
Why not do this in `Schedules::__init__`?  Then it wouldn't need to be a loop and schedule implementation quirks wouldn't leak to common rbd_support setup code.

idryomov replied:
> Initially, I wanted to add this in Schedules::**init**. But I didn't do that since the Schedule's init gets called too during scheduler handler's load().

Ah, yeah, I keep forgetting about that...

> But I figured that my current change will be simple enough for now, and easy to remove in a later version when we no longer need to keep doing this.

I don't like how this implementation detail is spilled into another file.  Have you considered adding static (class) attribute so it's done only once but still in `Schedules`?

ajarr replied:
I could only think of implementing it as a static method. Let me know if the latest version is better. If the static method doesn't look good then we can just move it out into just a plain function in schedule.py . Thanks!

ajarr replied:
No longer need to do this. Taking the approach describe here, https://github.com/ceph/ceph/pull/49261#discussion_r1046665725

trociny commented on src/pybind/mgr/rbd_support/mirror_snapshot_schedule.py:
```diff
@@ -369,6 +369,9 @@ def init_schedule_queue(self) -> None:
         self.images: Dict[str, Dict[str, Dict[str, str]]] = {}
+        # Previous versions incorrectly stored the global config in
+        # the localized module option. Check the config is here and fix it.
+        Schedules.replace_local_with_global_mod_option(self)
```
It does not look very nice to me. I would prefer if we had the schedules initialization here (`self.schedules =  Schedules(self)`), where this would be called. And then in `load_schedules` just `self.schedules.load(...)`. And the schedules' `load` function might need updating to reset the schedules before loading the new ones.

ajarr replied:
sounds good. I'll try this out.

ajarr replied:
Done. Let me know if it looks okay.

trociny commented on src/pybind/mgr/rbd_support/schedule.py:
```diff
@@ -377,28 +377,30 @@ def __init__(self, handler: Any) -> None:
         self.schedules: Dict[str, Schedule] = {}
+        # Previous versions incorrectly stored the global config in
+        # the localized module option. Check the config is here and fix it.
+        self.replace_local_with_global_module_option()
```
nit: I don't see much point in a function that is going to be used only once, the inline code would look good to me. But I am fine with the current code too.

trociny reviewed (approved):
LGTM if testing confirms it works.

idryomov commented on src/pybind/mgr/rbd_support/schedule.py:
```diff
@@ -399,6 +391,16 @@ def load(self,
+             namespace_validator: Optional[Callable] = None,
+             image_validator: Optional[Callable] = None) -> None:
+        self.level_specs = {}
+        self.schedules = {}
```
I'd add a line break here to make the three separate steps (1. reset 2. ingest global spec 3. ingest pool and image-level specs) clear.

idryomov replied:
It looks like you have also added a extra line break:
```
        schedule_cfg = self.handler.module.get_module_option(
            self.handler.MODULE_OPTION_NAME, '')

        if schedule_cfg:
```
which isn't needed because these lines are part of the same "2. injest global spec" step ;)

idryomov replied:
(I noticed only because I looked at the incremental diff instead of whole thing, feel free to ignore...)

ajarr replied:
Thanks! Fixed.

idryomov reviewed (commented):
You may want to drop `Co-authored-by` as there is no trace of Mykola's patch left here ;)

ajarr commented:
> You may want to drop `Co-authored-by` as there is no trace of Mykola's patch left here ;)

Removed it. I'd decided to keep it since Mykola spelled out the code for me here https://github.com/ceph/ceph/pull/49261#discussion_r1046665725 .  I'm good with both. :-)

idryomov commented:
https://pulpito.ceph.com/dis-2022-12-15_11:30:57-rbd-wip-dis-testing-distro-default-smithi/

idryomov commented:
jenkins test api

idryomov commented:
jenkins test api

idryomov commented:
jenkins test api

idryomov commented:
jenkins test windows

idryomov commented:
@petrutlucian94 Looks like the new "Fs" tests need some retries as well:
```
+ timeout 5m ssh -i **** -o UserKnownHostsFile=/home/jenkins-build/build/workspace/ceph-windows-pull-requests/libvirt/known_hosts administrator@192.168.122.164 python.exe /workspace/test_rbd_wnbd.py --test-name RbdFsTest --iterations 4
[2022-12-26 13:40:06,708] ERROR - Command failed: ('powershell.exe', '-command', 'Get-Disk -Number 4 | Initialize-Disk -PassThru | New-Partition -AssignDriveLetter -UseMaximumSize | Format-Volume -Force -Confirm:$false'). Return code: 1. Stdout: b''. Stderr: b'Exception calling "EndProcessing" with "0" argument(s): "Win32 internal error "Access is denied" 0x5 occurred while \r\nreading the console output buffer. Contact Microsoft Customer Support Services."\r\nAt line:695 char:17\r\n+                 $__cmdletization_objectModelWrapper.EndProcessing()\r\n+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\r\n    + CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordException\r\n    + FullyQualifiedErrorId : HostException\r\n \r\n'.
```

idryomov commented:
No related failures:

https://pulpito.ceph.com/dis-2022-12-21_08:46:24-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2022-12-21_21:28:24-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2022-12-23_14:36:40-rbd-wip-dis-testing-distro-default-smithi/
